snuba/web/views.py

```diff
@@ -269,46 +264,60 @@ def dataset_query(dataset: Dataset, body, timer: Timer) -> Response:
     assert http_request.method == "POST"
     ensure_not_internal(dataset)
     ensure_table_exists(dataset)
-    return format_result(
-        run_query(
-            dataset,
-            validate_request_content(
-                body,
-                RequestSchema.build_with_extensions(
-                    dataset.get_extensions(), HTTPRequestSettings
-                ),
-                timer,
-                dataset,
-                http_request.referrer,
-            ),
-            timer,
-        )
-    )
 
+    request = validate_request_content(
+        body,
+        RequestSchema.build_with_extensions(
+            dataset.get_extensions(), HTTPRequestSettings
+        ),
+        timer,
+        dataset,
+        http_request.referrer,
+    )
 
-def run_query(dataset: Dataset, request: Request, timer: Timer) -> WebQueryResult:
     try:
         result = parse_and_run_query(dataset, request, timer)
-        payload = {**result.result, "timing": timer.for_json()}
-        if settings.STATS_IN_RESPONSE or request.settings.get_debug():
-            payload.update(result.extra)
-        return WebQueryResult(payload, 200)
-    except RawQueryException as e:
-        return WebQueryResult(
-            {
-                "error": {"type": e.err_type, "message": e.message, **e.meta},
-                "sql": e.sql,
-                "stats": e.stats,
-                "timing": timer.for_json(),
-            },
-            429 if e.err_type == "rate-limited" else 500,
+    except RawQueryException as exception:
+        cause = getattr(exception, "__cause__", None)
+        details: Mapping[str, Any]
+        if isinstance(cause, RateLimitExceeded):
+            details = {
+                "type": "rate-limited",
+                "message": "rate limit exceeded",
+            }
+        elif isinstance(cause, ClickhouseError):
+            details = {
+                "type": "clickhouse",
+                "message": str(cause),
+                "code": cause.code,
+            }
+        elif isinstance(cause, Exception):
+            details = {
+                "type": "unknown",
+                "message": str(cause),
+            }
+        else:
+            raise NotImplementedError  # something went very wrong
+
+        return Response(
+            json.dumps(
+                {
+                    "error": details,
+                    "sql": exception.sql,
+                    "stats": exception.stats,
```

Comment:
Technically this is the same data that is in the `if settings.STATS_IN_RESPONSE or request.settings.get_debug()` conditional below on line 317 but it's never worked that way, which has always bugged me. I'm not sure if making it consistent at this point would break anything.

It might make sense to encapsulate that extra data in a type of it's own so it's clear that the only difference between `RawQueryResult` and `RawQueryException` is the presence of an actual `Result`, everything else is the same.

Reply:
Went ahead and pulled in that type consolidation change but not going to change the output format here.